Readme refactor (#93)
careem/mockingbird: README.md

MockingBird doesn't use any annotation processor or reflection. This means that it is a bit more verbose with respect to
libraries like `Mockito` or `Mockk`

### Mock generation plugin ( experimental )

The mock generation plugin generates mock boilerplate code for you, the plugin can be used along with manual mocks,
it is currently experimental and has several limitations.

> To use this plugin you have to use mockingbird version `2.0.0-beta04` or above, to see examples checkout `generate-mocks`

Reviewer: Can I suggest 2.0.0-beta06

Reviewer: ```suggestion
To use this plugin you have to use mockingbird version `2.0.0-beta06` or above, to see examples checkout `generate-mocks` branch
```

> and explore the `samples` folder, `samples` is a project itself, you can open `samples` as a standalone project.

Reviewer: ```suggestion
and explore the `samples` project, You can open `samples` is a standalone project.
```

NOTE: the plugin doesn't discover which interfaces to mock, it's up to you to configure those.

WARNING: If you do not what to use the plugin you can use the old way of manual mock generation
check [Mocks](https://github.com/careem/mockingbird#mocks) or [Spies](https://github.com/careem/mockingbird#spies)

#### Plugin Setup

To start using the plugin you need to include it in your project `build.gradle.kts` or `build.gradle` adding the
following line

```kotlin

> classpath("com.careem.mockingbird:mockingbird-compiler:2.0.0-beta05")

Reviewer: ```suggestion
classpath("com.careem.mockingbird:mockingbird-compiler:2.0.0-beta06")
```

Reviewer: Should we update this version as part of the release pipeline?

Author: replaced with variable so we do not need to update

```

be also sure you have `mavenCentral` in your `buildscripts` repositories, your project `build.gradle.kts` might look
similar the one below

```kotlin
buildscript {
    repositories {
        ...
        mavenCentral()
    }
    dependencies {
        ...

>         classpath("com.careem.mockingbird:mockingbird-compiler:2.0.0-beta05")

Reviewer: ```suggestion
        classpath("com.careem.mockingbird:mockingbird-compiler:2.0.0-beta06")
```

Author: same as above

    }
}
```

> To generate mocks for a specific module you have first to apply the plugin in your module `build.gradle.kts`

Reviewer: ```suggestion
To generate mocks for a specific module you have first to apply the plugin in your module
```

> ```kotlin

Reviewer: Groove DSL:

```suggestion
```kotlin
```

> apply(plugin = "com.careem.mockingbird")

Reviewer: can you move this under the kotlin section? and also put the groovy script since this is only for kotlin dsl

Author: Is ti worth tho? I specified Colton vs groovy only for the part where you specify the classes because it is different but for apply plugin I think we do not need to explain both no?

```

And then specify for what interfaces you what to generate mocks for, see the example below

> kotlin

Reviewer: ```suggestion
Kotlin DSL:
```

```kotlin
configure<com.careem.mockingbird.MockingbirdPluginExtension> {
    generateMocksFor = listOf(
        "com.careem.mockingbird.sample.Mock1",
        "com.careem.mockingbird.sample.MockWithExternalDependencies",
        "com.careem.mockingbird.common.sample.ExternalContract"
    )
}
```

> groovy

Reviewer: ```suggestion
Groovy DSL:
```

```groovy
mockingBird {
    generateMocksFor = [
            'com.careem.mockingbird.sample.Mock1',
            'com.careem.mockingbird.sample.MockWithExternalDependencies',
            'com.careem.mockingbird.common.sample.ExternalContract'
    ]
}
```

#### Plugin Usage

To generate mocks you can simply run `./gradlew generateMocks` or simply run `./gradlew build`, for farted development
loop using
`generatedMocks` is recommended

#### Plugin Limitations

* The plugin can only be used in modules containing a `jvm` target
* The plugin can generate mocks only, no support for spies yet
* You can mock interfaces only
* You cannot mock interfaces that have generic types in their definitions
* You cannot mock lambdas

> * You cannot mock suspend functions

Reviewer: can you also mention inline and reified?

Author: If I have to put everything this list will be endless XD

Author: added anyway

MockingBird doesn't use any annotation processor or reflection. This means that it is a bit more verbose with respect to
libraries like `Mockito` or `Mockk`

> ### Mock generation plugin ( experimental )

Reviewer 2: ```suggestion
### Mock generation plugin (experimental)
```

The mock generation plugin generates mock boilerplate code for you, the plugin can be used along with manual mocks, it
is currently experimental and has several limitations.

To use this plugin you have to use mockingbird version `2.0.0-beta06` or above, to see examples
checkout `generate-mocks` branch and explore the `samples` project, You can open `samples` is a standalone project.

NOTE: the plugin doesn't discover which interfaces to mock, it's up to you to configure those.

> WARNING: If you do not what to use the plugin you can use the old way of manual mock generation

Reviewer 2: ```suggestion
WARNING: If you do not what to use the plugin you can use the old way of manual mock generation,
```

check [Mocks](https://github.com/careem/mockingbird#mocks) or [Spies](https://github.com/careem/mockingbird#spies)

#### Plugin Setup

To start using the plugin you need to include it in your project `build.gradle.kts` or `build.gradle`

Be sure you have `mavenCentral()` in your `buildscripts` repositories, your project build gradle might look similar the
one below

Kotlin DSL:

```kotlin
buildscript {
    repositories {
        ...
        mavenCentral()
    }
    dependencies {
        ...
        classpath("com.careem.mockingbird:mockingbird-compiler:$mockingBirdVersion")
    }
}
```

Groovy DSL:

```groovy
buildscript {
    repositories {
        ...
        mavenCentral()
    }
    dependencies {
        ...
        classpath "com.careem.mockingbird:mockingbird-compiler:$mockingBirdVersion"
    }
}
```

> To generate mocks for a specific module you have first to apply the plugin in your module's build gradle

Reviewer 2: ```suggestion
To generate mocks for a specific module you have first to apply the plugin in this module's build gradle
```

Kotlin DSL:

```kotlin
configure<com.careem.mockingbird.MockingbirdPluginExtension> {
    generateMocksFor = listOf(
        "com.careem.mockingbird.sample.Mock1",
        "com.careem.mockingbird.sample.MockWithExternalDependencies",
        "com.careem.mockingbird.common.sample.ExternalContract"
    )
}
```

Groovy DSL

```groovy
mockingBird {
    generateMocksFor = [
            'com.careem.mockingbird.sample.Mock1',
            'com.careem.mockingbird.sample.MockWithExternalDependencies',
            'com.careem.mockingbird.common.sample.ExternalContract'
    ]
}
```

#### Plugin Usage

To generate mocks you can simply run `./gradlew generateMocks` or simply run `./gradlew build`, for faster development
loop using `generatedMocks` is recommended

#### Plugin Limitations

* The plugin can only be used in modules containing a `jvm` target
* The plugin can generate mocks only, no support for spies yet
* You can mock interfaces only
* You cannot mock interfaces that have generic types in their definitions
* You cannot mock lambdas
* You cannot mock suspend functions
* You cannot mock inline functions

> * You cannot mock reified functions

Reviewer 2: Nit: I'd suggest to make these passive, like `Only interfaces can be mocked`, `Generic interfaces cannot be mocked` etc.

    }
```

### Manual Mocks and Spies generation

In this section it is explained how to generate mocks and spy manually, we suggest to use this approach only when you
face issues with the Mock generator plugin.

#### Mocks

The first step you need to do is create a mock class for the object you want to mock, you need a mock for each
dependency type you want to mock

The library provides 2 functions to help you write your mocks.

> 1. `mock` this function allows you to mock non-Unit methods

Reviewer 2: ```suggestion
1. `mock` this function allows you to mock non-methods with return types other than Unit
```

    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and

>     limitations under the License.

Reviewer 2: What's happened here exactly? 🤔
